dcpy/utils/postgres.py

```diff
@@ -7,42 +7,143 @@
 from sqlalchemy import create_engine, text
 from typing import Optional
 
+DEFAULT_POSTGRES_SCHEMA = "public"
 
-BUILD_ENGINE_RAW = os.environ["BUILD_ENGINE"]
-build_engine = create_engine(BUILD_ENGINE_RAW)
 
-
-def execute_query(query: str, placeholders: Optional[dict] = None) -> None:
-    if placeholders is None:
-        placeholders = {}
-    with build_engine.connect() as connection:
-        connection.execute(statement=text(query), parameters=placeholders)
-
-
-def execute_select_query(
-    query: str, placeholders: Optional[dict] = None
-) -> pd.DataFrame:
-    if placeholders is None:
-        placeholders = {}
-    with build_engine.connect() as sql_conn:
-        select_records = pd.read_sql(sql=text(query), con=sql_conn, params=placeholders)
-    return select_records
+def generate_engine_uri(
+    server_url: str, database: str, schema: str = DEFAULT_POSTGRES_SCHEMA
+):
+    # the default postgres schema must always be in the search_path
+    schemas = (
+        f"{schema},{DEFAULT_POSTGRES_SCHEMA}"
+        if schema != DEFAULT_POSTGRES_SCHEMA
+        else schema
+    )
+    options = f"?options=--search_path%3D{schemas}"
+    return server_url + "/" + database + options
 
 
-def execute_file_via_shell(build_engine: str, path: Path):
+def execute_file_via_shell(engine_uri: str, path: Path):
     """Execute .sql script at given path."""
-    cmd = f"psql {build_engine} -v ON_ERROR_STOP=1 -f {path}"
+    cmd = f"psql {engine_uri} -v ON_ERROR_STOP=1 -f {path}"
     if os.system(cmd) != 0:
         raise Exception(f"{path} has errors!")
 
 
-def execute_via_shell(build_engine: str, sql_statement):
+def execute_query_via_shell(engine_uri: str, sql_statement):
     """Execute sql via psql shell."""
-    cmd = f"psql {build_engine} -v ON_ERROR_STOP=1 {sql_statement}"
+    cmd = f"psql {engine_uri} -v ON_ERROR_STOP=1 {sql_statement}"
     if os.system(cmd) != 0:
         raise Exception(f"Command has errors! {cmd}")
 
 
+class PostgresClient:
+    def __init__(
+        self,
+        schema: str,
+        *,
+        database: Optional[str] = None,
+        server_url: Optional[str] = None,
+    ):
+        self.schema = schema
+        self.database = database if database else os.environ["BUILD_ENGINE_DB"]
+        self.server_url = (
+            server_url if server_url else os.environ["BUILD_ENGINE_SERVER"]
+        )
+
+        self.engine_uri = generate_engine_uri(self.server_url, self.database, schema)
```

Comment:
@damonmcc It's a little confusing to have both an `engine_uri` and a `server_url` attribute. Should we just inline this `generate_engine_uri` call?

Reply:
true. having an `engine_uri` attribute seems more useful than `server_url`, like if a build creates a PostgreClient but then has to execute sql via shell, it could look like `execute_file_via_shell(pg_client.engine_uri, file_path)`

so maybe something like this then:
```python
self.engine_uri = generate_engine_uri(
    server_url if server_url else os.environ["BUILD_ENGINE_SERVER"],
    self.database,
    schema
)
self.engine = create_engine(
    self.engine_uri,
    isolation_level="AUTOCOMMIT",
)
```

Reply:
see [8f15ee3](https://github.com/NYCPlanning/data-engineering/pull/234/commits/8f15ee38f341a8005dbab3baa4ced2ade3922c9c)